### PR TITLE
more rewrites of the plugin

### PR DESCRIPTION
**Pull request by iionly**
@jeabakker If you would be interested to look at this PR, it would help me a lot to hear your opinion.

Two major changes: userpoints for friending and userpoints for annotations to allow for deletion of userpoints in both of these cases. Beside that it's some follow up stuff I noticed in your PR that I thought needed some polishing up.

The main problem with friending was that the userpoint entity description only contained info on the display name, so deletion of userpoints on unfriending was a major issue. I also switched to the relationship, create/delete hooks to be able to check existence of a userpoint entity for friending for the user pair in question. What I have not yet implemented is an update script for converting any existing friending userpoint entities to the new scheme - and I'm not sure I should do so or just leave any existing userpoint entities as they are. Problem here is that the conversion can't be unique in all cases (a user being friend with more than one other user who has the same display name or a user might have changed the display name in between). I don't even know if the new way is any better (i.e. I might need to alter it to save also (or instead) the user guid) because the username isn't necessarily permanent either (Profile Manager plugin).

With annotations the problem is for some annotations userpoints must not be awarded more than once (e.g. likes, voting on polls, fivestar ratings). But the userpoints awarded on creation of these annotations were not deleted in the past. One way to handle possible duplicates would be to leave them in the database them until a delete event for one of these annotations is fired and only then remove them all. Another way would be to just leave them and don't bother. Also, the definition of which types of annotations are "one-time awarding" cases is hard-coded for now. This is even more of a problem now as you made it possible to define additional userpoints actions by extending the actions form.

### CODE REVIEW

File: lib/userpoint.php [hunk 1]
```diff
@@ -192,15 +195,14 @@ function elggx_userpoints_get($user_guid) {
 		'pending' => 0,
 	];
 
-	$entities = elgg_get_entities_from_metadata([
+	$entities = new ElggBatch('elgg_get_entities_from_metadata', [
```
Comment (jeabakker) [inline lib/userpoint.php]: why did you change this, the result is exactly the same

File: lib/userpoint.php [hunk 2]
```diff
@@ -237,25 +298,30 @@ function elggx_userpoints_delete($user_guid, $guid) {
 	$access_status = access_get_show_hidden_status();
 	access_show_hidden_entities(true);
 
-	$points = 0;
+	$points_array = [];
 
-	$entities = elgg_get_entities_from_metadata([
-		'metadata_name' => 'meta_guid',
-		'metadata_value' => $guid,
+	$entities = new ElggBatch('elgg_get_entities_from_metadata', [
```
Comment (jeabakker) [inline lib/userpoint.php]: same question about the batch change

File: views/default/elggx_userpoints/list.php [hunk 3]
```diff
@@ -44,7 +44,7 @@
 		
 		// name
 		$row[] = elgg_format_element('td', ['width' => '50%'], elgg_view('output/url', [
-			'text' => $entity->getDisplayName(),
+			'text' => $entity->username,
```
Comment (jeabakker) [inline views/default/elggx_userpoints/list.php]: Why change this to username, and not keep the display name? If my username is `xwsdt` but by display name is `Jerome Bakker` i as an admin would prefer the display name. (or maybe `Jerome Bakker (xwsdt)`

File: actions/restore.php [hunk 4]
```diff
@@ -17,4 +17,4 @@
 $users_approved_points = $users_points['approved'];
 $user->userpoints_points = (int) $users_approved_points;
 
-return elgg_ok_response('', elgg_echo('elggx_userpoints:restore:success', [$user->getDisplayName()]), REFERER);
+return elgg_ok_response('', elgg_echo('elggx_userpoints:restore:success', [$username]), REFERER);
```
Comment (jeabakker) [inline actions/restore.php]: why changed this

File: lib/userpoint.php (hunk 1 above)
Comment (iionly) [reply lib/userpoint.php]: I thought it would be better to define the ElggBatches in the same way throughout the whole plugin. As the "traditional" approach was still used more often than the new one I was just lazy and changed the code back at the two places the new way was used.

Also, I wasn't 100% sure about ```setIncrementOffset()``` usage, i.e. if it's set to false automatically if the 'batch' option is set to true or if I would have to set it to false as before (it's necessary in the second location when deleting entities is done in the loop but you had removed the line before the loop). I guess I just could have asked to make sure that my feeling is right that it works in the same way as before and I would have to set it to false explicitely.

Actually, I hadn't known about the new 'batch' option for the e_get*() functions and I like it. As the PR is not ready anyway I can switch all ElggBatches definitions to the new approach.

File: lib/userpoint.php (hunk 2 above)
Comment (iionly) [reply lib/userpoint.php]: Same answer as above.

File: actions/restore.php (hunk 4 above)
Comment (iionly) [reply actions/restore.php]: The workflow here is to enter the username of the account you want to restore the userpoints count. Therefore, I thought it would be more consistent to use the username also in the success message you get - even more so as you still had the username in the error message in the action. If you would restore the points of only one account at a time it probably wouldn't matter to use the display name in the messages. But if it happens that you might need to restore a few at once it might be a distraction if you get reported a different name in the messages than you entered in the form.

File: views/default/elggx_userpoints/list.php (hunk 3 above)
Comment (iionly) [reply views/default/elggx_userpoints/list.php]: See also the answer below.

This list is on the same page as the restore input field where you need to enter the username. I thought I would keep it simple and only display the username on the whole page. As it's the admin backend I hadn't thought it would matter because no normal user would see that their username and not their display name is displayed (and if necessary worked with).

Showing both username (necessary to work with) and display name (for completeness) might be an option. Even more so as the details tab still uses the display name.

File: lib/userpoint.php (hunk 1 above)
Comment (iionly) [reply lib/userpoint.php]: Found the 'batch_inc_offset' option is available. That answers my question.
